### PR TITLE
Initial suggestion for a feign <-> resilience4j module

### PR DESCRIPTION
**Pull request by pwlan**
There is still a bit to do, but i wanted to incorporate your feedback early.

### CODE REVIEW

**Comment by RobWin** [comment]
Thx Pat.

**Comment by RobWin** [comment]
Could you please use Assertj instead of Junit assertions.

**Comment by pwlan** [comment]
Sure

File: resilience4j-feign/src/main/java/io/github/resilience4j/feign/DecoratorInvocationHandler.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    private final Target<?> target;
+    private final Map<Method, CheckedFunction1<Object[], Object>> decoratedDispatch;
+
+    public DecoratorInvocationHandler(Target<?> target, Map<Method, MethodHandler> dispatch, FeignDecorator invocationDecorator) {
+        checkNotNull(dispatch, "dispatch");
+        decoratedDispatch = decorateMethodHandlers(dispatch, invocationDecorator);
+        this.target = checkNotNull(target, "target");
+    }
+
+    private Map<Method, CheckedFunction1<Object[], Object>> decorateMethodHandlers(Map<Method, MethodHandler> dispatch,
```
Comment (RobWin) [inline resilience4j-feign/src/main/java/io/github/resilience4j/feign/DecoratorInvocationHandler.java]: Could you please add some javadoc at this method

File: resilience4j-feign/src/main/java/io/github/resilience4j/feign/FeignDecorators.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.ratelimiter.RateLimiter;
+import io.vavr.CheckedFunction1;
+
+/**
+ * Builder to help build stacked decorators.
+ */
+public class FeignDecorators implements FeignDecorator {
```
Comment (RobWin) [inline resilience4j-feign/src/main/java/io/github/resilience4j/feign/FeignDecorators.java]: Maybe we can add here that the order of the `withXXX` builder methods als controls the order of the decorators.

Comment (pwlan) [reply resilience4j-feign/src/main/java/io/github/resilience4j/feign/FeignDecorators.java]: Yes, I have been thinking about the ordering of the decorators. The question is; would it be better to have a static ordering? For example, does it make sense to have the Ratelimiter before the CircuitBreaker a.s.o.?

Comment (RobWin) [reply resilience4j-feign/src/main/java/io/github/resilience4j/feign/FeignDecorators.java]: I think the flexibility should be given

**Comment by pwlan** [comment]
@RobWin Ready for another review

**Comment by RobWin** [comment]
@storozhukBM Would you like to review it as well?

**Comment by RobWin** [comment]
@pwlan could you please rebase the latest changes from master?

**Comment by RobWin** [comment]
There seems to be an error in JavaDoc. The build failed on TraviscI.

**Comment by pwlan** [comment]
Build is fixed

**Comment by pwlan** [comment]
@RobWin Anything else to do here?

**Comment by RobWin** [comment]
No, but I'm not sure if we want to introduce support for Feign when the project itself is not maintained anymore.
See: https://github.com/OpenFeign/feign/issues/646

**Comment by fbyrne** [comment]
Hi, 

Any update on this request?  Seems more contributors have volunteered for feign.

Kind regards, 
Fergus

**Comment by storozhukBM** [comment]
Hi @pwlan,
If you're still interested in this PR, maybe we can make it through?
I think all we need to do here is merge with current resilience4j master, update feign version and fix possible bugs.

**Comment by pwlan** [comment]
@storozhukBM Sure, i will get on it on the weekend.

**Comment by storozhukBM** [comment]
Hi @pwlan thank you for this effort. Everything looks good here. Sorry for delayed response. We are going to make v0.13.2 with some bug-fixes and small improvements on the next week I think. And then I'll merge  your changes to be a part of v0.14.0

**Comment by storozhukBM** [comment]
@pwlan sorry for such a long wait on this. But now we've release v0.13.2 and we're working on v0.14.0

**Comment by storozhukBM** [comment]
@pwlan Thank you once again for this contribution

**Pull request merged**